### PR TITLE
RPD internal air tank + RPD increased reach

### PR DESCRIPTION
**Pull request by Ghilker**
## About The Pull Request
The RPD now function thanks to an internal air tank that releases the gases on use.
There is a setting that allows the players to increase the reach of the RPD up to 4 tiles away at the cost of increased gas consumption.
Here can be seen the various tiles the RPD can now reach from the central position (each color indicates the furthest that the setting allows)
image

## Why It's Good For The Game
A more balanced device and an increase in tool quality for the players by allowing easier setup making.

## Changelog

:cl:
expansion: RPD consumes gases from an internal tank when used, the reach the device can have can be changed to up 4 tiles away at the cost of increased consumption (Note that even at max reach, placing pipes closer will lower the consumption to the appropriate amount)
/:cl:

### CODE REVIEW

**Comment by bruhlookatthisdood** [comment]
how do refill?

**Comment by bruhlookatthisdood** [comment]
> > how do refill?
> 
> you can remove the tank with alt click (is all explained in the examine messages) and replace it

nice, will it be able to go through windows?
also what gas does it take, just normal oxygen? and if so is it just like the flamethrower except you install a oxygen tank instead of a plasma tank?

**Comment by Ghilker** [comment]
> nice, will it be able to go through windows? also what gas does it take, just normal oxygen? and if so is it just like the flamethrower except you install a oxygen tank instead of a plasma tank?

It should be able to go over windows yes
It takes virtually any gas and is as easy as to use a flamethrower yes. Just remember to use an inert or non-harmful gas since it spews the gases in the turf on use

**Comment by MacBlaze1** [comment]
> > it spews the gases in the turf on use
> 
> oooh i like that feature seems like an easy way to plasma flood a room though. or... whatever other gas is inside it _cough_ _cough_ **fusion** _cough_ _cough_
I dont think it will release enough to flood a room

**Comment by derekhearst** [comment]
Can this dispense through windows and other tiles? Does this change disposals and transit tubes too?

File: code/game/objects/items/RPD.dm [hunk 1]
```diff
@@ -237,10 +237,15 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
 	var/obj/item/tank/internal_tank
+	///Differentiate the normal RPD from the borg ones since they don't use pressure
+	var/use_pressure = TRUE
 
 /obj/item/pipe_dispenser/constructed
 	constructed = TRUE
 
+/obj/item/pipe_dispenser/cyborg
+	use_pressure = FALSE
+
```
Comment (zxaber) [inline code/game/objects/items/RPD.dm:248]: Assuming this is due to refil concerns, you could also have attacking a canister with the RPD work as filling it up (perhaps with a do-after simulating attaching the RPD's quick-refil nozzle to the canister's tank port)

Then the borg RPD could work identically to the crew's version.

Comment (MacBlaze1) [reply code/game/objects/items/RPD.dm:248]: engineering borgs taking more of our jobs is a bad thing

**Review by MrMelbert — CHANGES_REQUESTED** [review]
i'm gonna put a "no" on this if you can use the RPD over walls, windows, and doors - long distance plasmaflooding with x-ray sounds incredibly un-fun

**Comment by Ghilker** [comment]
> i'm gonna put a "no" on this if you can use the RPD over walls, windows, and doors - long distance plasmaflooding with x-ray sounds incredibly un-fun

aight willl look into it

**Comment by Ghilker** [comment]
RPD now can't place devices over walls/windows/doors, also fixed an issue where you could attack stuff at range

**Comment by bruhlookatthisdood** [comment]
wait, will it still cost gas at one tile range? (right next to you)

**Comment by Ghilker** [comment]
yes

File: code/game/objects/items/RPD.dm [hunk 2]
```diff
@@ -420,7 +473,19 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	if(!ISADVANCEDTOOLUSER(user) || istype(A, /turf/open/space/transit))
 		return ..()
 
+	if(!internal_tank)
+		balloon_alert(user, "Missing air tank")
```
Comment (MrMelbert) [inline code/game/objects/items/RPD.dm:477]:
```suggestion
		balloon_alert(user, "missing air tank!")
```

File: code/game/objects/items/RPD.dm [hunk 3]
```diff
@@ -429,6 +494,11 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 
 	//Unwrench pipe before we build one over/paint it, but only if we're not already running a do_after on it already to prevent a potential runtime.
 	if((mode & DESTROY_MODE) && (upgrade_flags & RPD_UPGRADE_UNWRENCH) && istype(attack_target, /obj/machinery/atmospherics) && !(DOING_INTERACTION_WITH_TARGET(user, attack_target)))
+		if(!consume_pressure(distance, full_consumption = FALSE))
+			balloon_alert(user, "not enough pressure")
```
Comment (MrMelbert) [inline code/game/objects/items/RPD.dm:498]:
```suggestion
			balloon_alert(user, "not enough pressure!")
```

File: code/game/objects/items/RPD.dm [hunk 4]
```diff
@@ -444,6 +514,11 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 	. = TRUE
 
 	if((mode & DESTROY_MODE) && istype(attack_target, /obj/item/pipe) || istype(attack_target, /obj/structure/disposalconstruct) || istype(attack_target, /obj/structure/c_transit_tube) || istype(attack_target, /obj/structure/c_transit_tube_pod) || istype(attack_target, /obj/item/pipe_meter) || istype(attack_target, /obj/structure/disposalpipe/broken))
+		if(!consume_pressure(distance, full_consumption = FALSE))
+			balloon_alert(user, "not enough pressure")
```
Comment (MrMelbert) [inline code/game/objects/items/RPD.dm:518]:
```suggestion
			balloon_alert(user, "not enough pressure!")
```

File: code/game/objects/items/RPD.dm [hunk 5]
```diff
@@ -523,9 +598,15 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
 			I.update()
 
 	if(mode & BUILD_MODE)
+		if(!consume_pressure(distance, full_consumption = TRUE))
+			balloon_alert(user, "not enough pressure")
```
Comment (MrMelbert) [inline code/game/objects/items/RPD.dm:602]:
```suggestion
			balloon_alert(user, "not enough pressure!")
```

File: code/game/objects/items/RPD.dm [hunk 6]
```diff
@@ -287,8 +319,29 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+	if(istype(W, /obj/item/tank))
+		if(internal_tank != null)
+			to_chat(user, span_notice("A tank is already inserted."))
+			return
+		if(!user.transferItemToLoc(W, src))
+			return
+		var/obj/item/tank/new_tank = W
+		internal_tank = new_tank
+		to_chat(user, span_notice("You put [W] in [src]."))
+		return
```
Comment (MrMelbert) [inline code/game/objects/items/RPD.dm:331]: These two can be balloon alerts

File: code/game/objects/items/RPD.dm [hunk 7]
```diff
@@ -240,15 +260,27 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+		qdel(internal_tank)
 	return ..()
 
 /obj/item/pipe_dispenser/examine(mob/user)
 	. = ..()
 	. += "You can scroll your mouse wheel to change the piping layer."
 	. += "You can right click a pipe to set the RPD to its color and layer."
+	. += "You can set the pressure strenght with Ctrl click, current strenght at [pressure_strenght]."
+	if(internal_tank)
+		. += "You can remove the internal tank with Alt click, current pressure at [internal_tank.air_contents.return_pressure()]"
```
Comment (MrMelbert) [inline code/game/objects/items/RPD.dm:281]:
```suggestion
		. += "You can remove the internal tank with Alt click, current pressure at [internal_tank.air_contents.return_pressure()]."
```

File: code/game/objects/items/RPD.dm [hunk 8]
```diff
@@ -240,15 +260,27 @@ GLOBAL_LIST_INIT(transit_tube_recipes, list(
...
+	else
+		. += "It's missing an air tank, please insert one to allow the device to work."
```
Comment (MrMelbert) [inline code/game/objects/items/RPD.dm:283]:
```suggestion
		. += "It's missing an air tank. Insert one to allow the device to work."
```

**Review by MrMelbert — COMMENTED** [review]
no longer blocking this, however I have very little to comment about the feature itself (Lemon may be more interested in it)

Here's some grammar and balloon alert tweaks

**Comment by Serxule** [comment]
also, maybe add a gauge like a welder, also how many pipes can you make before needing to refill?

**Comment by Ghilker** [comment]
> also, maybe add a gauge like a welder, also how many pipes can you make before needing to refill?

You can examine it to know how much pressure you've got.
If you place all the pipes close to you you will have a total of 260 pipes

**Comment by LemonInTheDark** [comment]
I'm not sure how much this will help, unless you already have a design built fully in your head. How much faster does this actually make things?

**Comment by Ghilker** [comment]
Placing pipes would be much faster since you don't have to move too much while working in a single room, like when piping atmos for a turbine burn.

Also it helps reducing errors (at least for me) when piping because less character movement allows more precise mouse positioning.

And it somewhat balances the RPD by requiring gas to operate (similar to how the rcd requires mats but less invasive and consuming)

**Comment by derekhearst** [comment]
How does this effect the cyborg rpd?

**Comment by Ghilker** [comment]
Rethinking...

**Comment by LemonInTheDark** [comment]
For the record, I'm not sure how useful this'll be, and it feels a bit uh, janky? if that makes sense. smells a bit off, and I don't really mind the rpd being infinite

**Comment by Ghilker** [comment]
> For the record, I'm not sure how useful this'll be, and it feels a bit uh, janky? if that makes sense. smells a bit off, and I don't really mind the rpd being infinite

Yeah after some thoughts It would have been fine to just have it on range, but the tank thing wasn't quite what I expected, don't worry about this one

**Pull request closed without merging**